File: py/tests/local/test_knowledgehub.py
"""
KnowledgeHub resource test — sync + async, runs against prodv2 gateway (data-board service).
Usage:
    pip install -e ".[dev]"
    python tests/local/test_knowledgehub.py

Notes:
  - data-board wraps all responses: { success, message, data } — SDK unwraps .data
  - createFolder requires organization_id, parent_folder_id, source_type (DB schema required)
  - getFolder returns { data: { folder, files } } — SDK unwraps to data["folder"]
  - delete_folders body uses { ids: [...] }
  - createFile needs actual file metadata (key, file_size, file_type) — skip in tests
"""
import os
import asyncio
from imbrace import ImbraceClient
from imbrace.async_client import AsyncImbraceClient

ACCESS_TOKEN = os.environ.get("IMBRACE_ACCESS_TOKEN", "acc_c8c27f3b-e147-4735-b641-61e8d3706692")
GATEWAY = os.environ.get("IMBRACE_GATEWAY_URL", "https://app-gatewayv2.imbrace.co")

passed = 0
failed = 0
skipped = 0


def ok(label, detail=""):
    global passed
    detail_str = f"  →  {str(detail)[:100]}" if detail else ""
    print(f"  ✓ {label}{detail_str}")
    passed += 1


def fail(label, err):
    global failed
    print(f"  ✗ {label}: {err}")
    failed += 1


def skip(label, reason):
    global skipped
    print(f"  - {label}  (skipped: {reason})")
    skipped += 1


# ── Sync Tests ────────────────────────────────────────────────────────────────

def test_sync():
    created = {"org_id": None, "folder_id": None, "sub_folder_id": None, "existing_folder_id": None, "existing_file_id": None}
    client = ImbraceClient(access_token=ACCESS_TOKEN, gateway=GATEWAY)
    boards = client.boards

    print("\n[1] Search folders (sync)")
    try:
        folders = boards.search_folders()
        if folders and isinstance(folders, list) and len(folders) > 0:
            created["org_id"] = folders[0].get("organization_id")
            created["existing_folder_id"] = folders[0].get("_id")
        ok("search_folders()", f"{len(folders)} folders  org={created['org_id']}")
    except Exception as e:
        fail("search_folders()", e)

    print("\n[2] Create folder (sync)")
    if not created["org_id"]:
        skip("create_folder()", "no org_id from search_folders")
    else:
        try:
            folder = boards.create_folder({
                "name": "SDK Python Test Folder",
                "organization_id": created["org_id"],
                "parent_folder_id": "root",
                "source_type": "upload",
            })
            created["folder_id"] = folder.get("_id")
            ok("create_folder()", f"id={created['folder_id']}")
        except Exception as e:
            fail("create_folder()", e)

    print("\n[3] Get folder (sync)")
    if not created["folder_id"]:
        skip("get_folder()", "no folder created")
    else:
        try:
            folder = boards.get_folder(created["folder_id"])
            ok("get_folder()", f"id={folder.get('_id')} name={folder.get('name')}")
        except Exception as e:
            fail("get_folder()", e)

    print("\n[4] Folder contents (sync)")
    if not created["folder_id"]:
        skip("get_folder_contents()", "no folder created")
    else:
        try:
            contents = boards.get_folder_contents(created["folder_id"])
            ok("get_folder_contents()", f"files={len(contents.get('files', []))} subfolders={len(contents.get('subfolders', []))}")
        except Exception as e:
            fail("get_folder_contents()", e)

    print("\n[5] Update folder (sync)")
    if not created["folder_id"]:
        skip("update_folder()", "no folder created")
    else:
        try:
            folder = boards.update_folder(created["folder_id"], {"name": "SDK Python Test Folder Updated"})
            ok("update_folder()", f"name={folder.get('name')}")
        except Exception as e:
            fail("update_folder()", e)

    print("\n[6] Create sub-folder (sync)")
    if not created["folder_id"] or not created["org_id"]:
        skip("create_folder() sub", "no parent folder")
    else:
        try:
            sub = boards.create_folder({
                "name": "SDK Python Test Sub-Folder",
                "organization_id": created["org_id"],
                "parent_folder_id": created["folder_id"],
                "source_type": "upload",
            })
            created["sub_folder_id"] = sub.get("_id")
            ok("create_folder() sub", f"id={created['sub_folder_id']}")
        except Exception as e:
            fail("create_folder() sub", e)

    print("\n[7] Search files (sync)")
    if not created["existing_folder_id"]:
        skip("search_files()", "no existing folder")
    else:
        try:
            files = boards.search_files(created["existing_folder_id"])
            if files and isinstance(files, list) and len(files) > 0:
                created["existing_file_id"] = files[0].get("_id")
            ok("search_files()", f"{len(files)} files")
        except Exception as e:
            fail("search_files()", e)

    print("\n[8] Get file (sync)")
    if not created["existing_file_id"]:
        skip("get_file()", "no file from search_files")
    else:
        try:
            f = boards.get_file(created["existing_file_id"])
            ok("get_file()", f"id={f.get('_id')} name={f.get('name')}")
        except Exception as e:
            fail("get_file()", e)

    print("\n[9] Cleanup folders (sync)")
    ids_to_delete = [x for x in [created["sub_folder_id"], created["folder_id"]] if x]
    if not ids_to_delete:
        skip("delete_folders()", "no folders created")
    else:
        try:
            res = boards.delete_folders(ids_to_delete)
            ok("delete_folders()", str(res)[:80])
            created["folder_id"] = None
            created["sub_folder_id"] = None
        except Exception as e:
            fail("delete_folders()", e)


# ── Async Tests ───────────────────────────────────────────────────────────────

async def test_async():
    created = {"org_id": None, "folder_id": None, "existing_folder_id": None, "existing_file_id": None}
    client = AsyncImbraceClient(access_token=ACCESS_TOKEN, gateway=GATEWAY)
    boards = client.boards

    print("\n[10] Search folders (async)")
    try:
        folders = await boards.search_folders()
        if folders and isinstance(folders, list) and len(folders) > 0:
            created["org_id"] = folders[0].get("organization_id")
            created["existing_folder_id"] = folders[0].get("_id")
        ok("async search_folders()", f"{len(folders)} folders")
    except Exception as e:
        fail("async search_folders()", e)

    print("\n[11] Create + Get + Delete folder (async)")
    if not created["org_id"]:
        skip("async folder CRUD", "no org_id")
    else:
        try:
            folder = await boards.create_folder({
                "name": "SDK Python Async Test Folder",
                "organization_id": created["org_id"],
                "parent_folder_id": "root",
                "source_type": "upload",
            })
            created["folder_id"] = folder.get("_id")
            ok("async create_folder()", f"id={created['folder_id']}")

            fetched = await boards.get_folder(created["folder_id"])
            ok("async get_folder()", f"id={fetched.get('_id')} name={fetched.get('name')}")

            contents = await boards.get_folder_contents(created["folder_id"])
            ok("async get_folder_contents()", f"files={len(contents.get('files', []))}")
        except Exception as e:
            fail("async folder CRUD", e)

    print("\n[12] Search + Get file (async)")
    if not created["existing_folder_id"]:
        skip("async file read", "no existing folder")
    else:
        try:
            files = await boards.search_files(created["existing_folder_id"])
            ok("async search_files()", f"{len(files)} files")
            if files:
                f = await boards.get_file(files[0].get("_id"))
                ok("async get_file()", f"id={f.get('_id')} name={f.get('name')}")
            else:
                skip("async get_file()", "no files in folder")
        except Exception as e:
            fail("async file read", e)

    print("\n[13] Async cleanup")
    if created["folder_id"]:
        try:
            res = await boards.delete_folders([created["folder_id"]])
            ok("async delete_folders()", str(res)[:60])
        except Exception as e:
            fail("async delete_folders()", e)

    await client.close()


# ── Main ──────────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    test_sync()
    asyncio.run(test_async())

    print(f"\n{'─' * 55}")
    print(f"  {passed} passed  |  {failed} failed  |  {skipped} skipped")
    if failed > 0:
        raise SystemExit(1)

File: ts/tests/local/test-knowledgehub.mjs
/**
 * KnowledgeHub resource test — runs against prodv2 gateway (data-board service).
 * Usage: node test-knowledgehub.mjs
 *
 * Routes through {gw}/data-board/folders and {gw}/data-board/files
 *
 * Notes:
 *  - data-board wraps all responses in { success, message, data } — SDK unwraps .data
 *  - createFolder requires organization_id, parent_folder_id, source_type (required by DB schema)
 *  - getFolder response is { data: { folder, files } } — SDK unwraps to data.folder
 *  - createFile requires actual file metadata (key, file_size, file_type) — use uploadFile instead
 *  - deleteFolders body uses { ids: [...] }
 */

import { ImbraceClient } from '@imbrace/sdk'

const ACCESS_TOKEN = process.env.IMBRACE_ACCESS_TOKEN || 'acc_c8c27f3b-e147-4735-b641-61e8d3706692'
const GATEWAY      = process.env.IMBRACE_GATEWAY_URL  || 'https://app-gatewayv2.imbrace.co'

const client = new ImbraceClient({ accessToken: ACCESS_TOKEN, baseUrl: GATEWAY })
const boards = client.boards

let passed = 0, failed = 0, skipped = 0
const created = {
  organizationId: null,
  folderId: null,
  subFolderId: null,
  existingFileId: null,   // from searchFiles on existing folder
  existingFolderId: null, // existing folder for file tests
}

function ok(label, detail = '') {
  console.log(`  ✓ ${label}${detail ? `  →  ${String(detail).slice(0, 100)}` : ''}`)
  passed++
}

function fail(label, err) {
  console.error(`  ✗ ${label}: ${err?.message ?? JSON.stringify(err)}`)
  failed++
}

function skip(label, reason) {
  console.log(`  - ${label}  (skipped: ${reason})`)
  skipped++
}

// ── [1] Search folders ────────────────────────────────────────────────────────

console.log('\n[1] Search folders')
try {
  const list = await boards.searchFolders({})
  const folders = Array.isArray(list) ? list : (list.data ?? [])
  if (folders.length > 0) {
    created.organizationId = folders[0].organization_id ?? null
    created.existingFolderId = folders[0]._id ?? folders[0].id
  }
  ok('searchFolders()', `${folders.length} folders  org=${created.organizationId}`)
} catch (e) { fail('searchFolders()', e) }

// ── [2] Create folder ─────────────────────────────────────────────────────────

console.log('\n[2] Create folder')
if (!created.organizationId) {
  skip('createFolder()', 'no organization_id from searchFolders')
} else {
  try {
    const folder = await boards.createFolder({
      name: 'SDK Test Folder',
      organization_id: created.organizationId,
      parent_folder_id: 'root',
      source_type: 'upload',
    })
    created.folderId = folder._id ?? folder.id
    ok('createFolder()', `id=${created.folderId}`)
  } catch (e) { fail('createFolder()', e) }
}

// ── [3] Get folder ────────────────────────────────────────────────────────────

console.log('\n[3] Get folder')
if (!created.folderId) {
  skip('getFolder()', 'no folder created')
} else {
  try {
    const folder = await boards.getFolder(created.folderId)
    ok('getFolder()', `id=${folder._id ?? folder.id} name=${folder.name}`)
  } catch (e) { fail('getFolder()', e) }
}

// ── [4] Folder contents ───────────────────────────────────────────────────────

console.log('\n[4] Folder contents')
if (!created.folderId) {
  skip('getFolderContents()', 'no folder created')
} else {
  try {
    const contents = await boards.getFolderContents(created.folderId)
    ok('getFolderContents()', `files=${contents.files?.length ?? 0} subfolders=${contents.subfolders?.length ?? 0}`)
  } catch (e) { fail('getFolderContents()', e) }
}

// ── [5] Update folder ─────────────────────────────────────────────────────────

console.log('\n[5] Update folder')
if (!created.folderId) {
  skip('updateFolder()', 'no folder created')
} else {
  try {
    const folder = await boards.updateFolder(created.folderId, { name: 'SDK Test Folder Updated' })
    ok('updateFolder()', `name=${folder.name}`)
  } catch (e) { fail('updateFolder()', e) }
}

// ── [6] Create sub-folder ─────────────────────────────────────────────────────

console.log('\n[6] Create sub-folder')
if (!created.folderId || !created.organizationId) {
  skip('createFolder() sub', 'no parent folder created')
} else {
  try {
    const sub = await boards.createFolder({
      name: 'SDK Test Sub-Folder',
      organization_id: created.organizationId,
      parent_folder_id: created.folderId,
      source_type: 'upload',
    })
    created.subFolderId = sub._id ?? sub.id
    ok('createFolder() sub', `id=${created.subFolderId}`)
  } catch (e) { fail('createFolder() sub', e) }
}

// ── [7] Search files in existing folder ──────────────────────────────────────

console.log('\n[7] Search files')
if (!created.existingFolderId) {
  skip('searchFiles()', 'no existing folder from search')
} else {
  try {
    const list = await boards.searchFiles({ folderId: created.existingFolderId })
    const files = Array.isArray(list) ? list : (list.data ?? [])
    if (files.length > 0) created.existingFileId = files[0]._id ?? files[0].id
    ok('searchFiles()', `${files.length} files`)
  } catch (e) { fail('searchFiles()', e) }
}

// ── [8] Get file ──────────────────────────────────────────────────────────────

console.log('\n[8] Get file')
if (!created.existingFileId) {
  skip('getFile()', 'no file from searchFiles')
} else {
  try {
    const file = await boards.getFile(created.existingFileId)
    ok('getFile()', `id=${file._id ?? file.id} name=${file.name}`)
  } catch (e) { fail('getFile()', e) }
}

// ── [9] Cleanup folders ───────────────────────────────────────────────────────

console.log('\n[9] Cleanup folders')
const foldersToDelete = [created.subFolderId, created.folderId].filter(Boolean)
if (foldersToDelete.length === 0) {
  skip('deleteFolders()', 'no folders created')
} else {
  try {
    const res = await boards.deleteFolders({ ids: foldersToDelete })
    ok('deleteFolders()', JSON.stringify(res).slice(0, 80))
    created.folderId = null
    created.subFolderId = null
  } catch (e) { fail('deleteFolders()', e) }
}

// ── Summary ───────────────────────────────────────────────────────────────────

console.log(`\n${'─'.repeat(50)}`)
console.log(`  ${passed} passed  |  ${failed} failed  |  ${skipped} skipped`)
if (failed > 0) process.exit(1)
